java/google/registry/rde/XjcToDomainResourceConverter.java

```diff
@@ -0,0 +1,329 @@
+          return convertSecdnsDsDataType(secdns);
+        }
+
+  };
+
+  /** Converts {@link XjcRgpStatus} to {@link GracePeriod} */
+  private static class GracePeriodConverter
+      implements Function<XjcRgpStatusType, GracePeriod> {
+
+    private final XjcRdeDomain domain;
+    private final Key<BillingEvent.Recurring> autoRenewBillingEvent;
+    private final Registry tld;
+
+    GracePeriodConverter(XjcRdeDomain domain, Key<BillingEvent.Recurring> autoRenewBillingEvent) {
+      this.domain = domain;
+      this.autoRenewBillingEvent = autoRenewBillingEvent;
+      this.tld = Registry.get(
+          Registries.findTldForNameOrThrow(InternetDomainName.from(domain.getName())).toString());
+    }
+
+    @Override
+    public GracePeriod apply(XjcRgpStatusType gracePeriodStatus) {
+      // TODO: (wolfgang) address these items in code review:
+      // verify that this logic is correct
+      switch(gracePeriodStatus.getS()) {
+        case ADD_PERIOD:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.ADD,
+              domain.getCrDate().plus(this.tld.getAddGracePeriodLength()),
+              domain.getCrRr().getClient());
+        case AUTO_RENEW_PERIOD:
+          return GracePeriod.createForRecurring(
+              GracePeriodStatus.AUTO_RENEW,
+              domain.getUpDate().plus(this.tld.getAutoRenewGracePeriodLength()),
+              domain.getClID(),
+              autoRenewBillingEvent);
+        case PENDING_DELETE:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.PENDING_DELETE,
+              domain.getUpDate().plus(this.tld.getPendingDeleteLength()),
+              domain.getClID());
+        case REDEMPTION_PERIOD:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.REDEMPTION,
+              domain.getUpDate().plus(this.tld.getRedemptionGracePeriodLength()),
+              domain.getClID());
+        case RENEW_PERIOD:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.RENEW,
+              domain.getUpDate().plus(this.tld.getRenewGracePeriodLength()),
+              domain.getClID());
+        case TRANSFER_PERIOD:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.TRANSFER,
+              domain.getUpDate().plus(this.tld.getTransferGracePeriodLength()),
+              domain.getTrnData().getReRr().getValue());
+        default:
+          throw new IllegalArgumentException(
+              "Unsupported grace period status: " + gracePeriodStatus.getS());
+      }
+    }
+  }
+
+  /** Converts {@link XjcRdeDomain} to {@link DomainResource}. */
+  static DomainResource convertDomain(XjcRdeDomain domain) {
+    // First create history entry and autorenew billing event/poll message
+    // Autorenew billing event is required for creating AUTO_RENEW grace period
+    HistoryEntry historyEntry = createHistoryEntry(domain);
```

Comment:
This approach seems wrong to me.  Each of these helper methods is individually running a separate transaction to create history entries, billing events, etc.  You want to save them all in a single transaction so that they all either succeed or fail.  Adjust these helpers to just create the entity rather than saving them, then run a single transaction at the end here to store all entities.  You don't want to end up in a situation where you persist some entities, and then hit a failure, and now the database is in a bad state with persisted history entries and billing events for a domain that itself doesn't exist.